src/EditorFeatures/Text/Implementation/TextBufferFactoryService/TextBufferCloneServiceFactory.cs

```diff
@@ -30,20 +31,42 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
         private class TextBufferCloneService : ITextBufferCloneService
         {
             private readonly ITextBufferFactoryService3 _textBufferFactoryService;
+            private readonly IContentType _roslynContentType;
             private readonly IContentType _unknownContentType;
 
             [ImportingConstructor]
             public TextBufferCloneService(ITextBufferFactoryService3 textBufferFactoryService, IContentTypeRegistryService contentTypeRegistryService)
             {
                 _textBufferFactoryService = textBufferFactoryService;
+
+                _roslynContentType = contentTypeRegistryService.GetContentType(ContentTypeNames.RoslynContentType);
                 _unknownContentType = contentTypeRegistryService.UnknownContentType;
             }
 
-            public ITextBuffer Clone(SnapshotSpan span)
+            public ITextBuffer CloneWithUnknownContentType(SnapshotSpan span)
                 => _textBufferFactoryService.CreateTextBuffer(span, _unknownContentType);
 
-            public ITextBuffer Clone(ITextImage textImage)
-                => _textBufferFactoryService.CreateTextBuffer(textImage, _unknownContentType);
+            public ITextBuffer CloneWithUnknownContentType(ITextImage textImage)
+                => Clone(textImage, _unknownContentType);
+
+            public ITextBuffer CloneWithRoslynContentType(SourceText sourceText)
+                => Clone(sourceText, _roslynContentType);
+
+            public ITextBuffer Clone(SourceText sourceText, IContentType contentType)
+            {
+                // see whether we can do it cheaply
+                var textImage = sourceText.TryFindCorrespondingEditorTextImage();
+                if (textImage != null)
+                {
+                    return Clone(textImage, contentType);
+                }
+
+                // we can't, so do it more expensive way
```

Comment:
expensive way was the way FAR used to do. I moved this code from there. and use cheap way if possible, which is almost 100% time for FAR.

if sourcetext is not backed by editor text (like sourcetext created from factory or modified tree from tree tranformation), this will still clone but through all expensive way.

Reply:
➡️ Discussion moved to #30868